File: backend/src/services/configurator/labourEngine.js
'use strict';

/**
 * labourEngine.js — pure functions, no I/O.
 *
 * The Python `quotation_calc.py` rolls labour up section-by-section
 * inside `compute_quote`. This module provides standalone helpers so
 * the labour calculation can be reused independently of pricing
 * (e.g. for the per-step labour preview in the configurator UI).
 *
 * It is mathematically identical to the labour block of `pricingEngine`
 * — same categories, same `hours = hours_per_unit * qty` formula, same
 * `cost = hours * rate` formula.
 */

const { LABOR_CATEGORIES } = require('./pricingEngine');

const num = (v, fallback = 0) => (typeof v === 'number' && Number.isFinite(v) ? v : fallback);

/**
 * Aggregate labour-hours per category from a flat BOM-row list.
 *
 * Each row may carry `lbr_cu`, `lbr_asm`, … `lbr_cad` properties (the
 * configurator component-master columns) plus `quantity`. A row with
 * `quantity=2, lbr_cu=0.5` contributes 1 hour to `cu`.
 *
 * Returns:
 *   {
 *     hours: { CU, ASM, CNT, QC, TST, ENG, CAD },   // floats
 *     totals: { hours_total }
 *   }
 */
function aggregateHoursFromBomRows(rows) {
  const hours = Object.fromEntries(LABOR_CATEGORIES.map((c) => [c, 0]));
  for (const row of rows || []) {
    const qty = num(row.quantity, 1);
    for (const cat of LABOR_CATEGORIES) {
      const colKey = `lbr_${cat.toLowerCase()}`;
      hours[cat] += num(row[colKey], 0) * qty;
    }
  }
  const hours_total = LABOR_CATEGORIES.reduce((acc, c) => acc + hours[c], 0);
  return { hours, totals: { hours_total } };
}

/**
 * Apply the LookupRates to an hours-per-category bag and return cost
 * per category plus the grand total. Symmetric with the labour block
 * inside Python `compute_quote`.
 */
function costFromHours(hours, lookup) {
  const result = { hours: {}, costs: {}, rates: {} };
  let costTotal = 0;
  let hoursTotal = 0;
  for (const cat of LABOR_CATEGORIES) {
    const h = num(hours[cat], 0);
    const rate = num(lookup[`LBR_${cat}_rate`], 0);
    const cost = h * rate;
    result.hours[cat] = h;
    result.costs[cat] = cost;
    result.rates[cat] = rate;
    hoursTotal += h;
    costTotal += cost;
  }
  result.totals = { hours_total: hoursTotal, cost_total: costTotal };
  return result;
}

/**
 * One-shot: BOM rows + LookupRates → labour costs.
 * Equivalent to `costFromHours(aggregateHoursFromBomRows(rows).hours, lookup)`.
 */
function computeLabour(rows, lookup) {
  const { hours } = aggregateHoursFromBomRows(rows);
  return costFromHours(hours, lookup);
}

module.exports = {
  LABOR_CATEGORIES,
  aggregateHoursFromBomRows,
  costFromHours,
  computeLabour,
};

File: java-backend/src/main/java/com/forge/configurator/engine/LabourEngine.java
package com.forge.configurator.engine;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class LabourEngine {

    public record LabourSummary(Map<String, Double> hours,
                                Map<String, Double> costs,
                                Map<String, Double> rates,
                                Map<String, Double> totals) {
    }

    public Map<String, Double> aggregateHoursFromBomRows(List<BomEngine.BomRow> rows) {
        Map<String, Double> hours = new LinkedHashMap<>();
        for (String cat : PricingEngine.LABOR_CATEGORIES) {
            hours.put(cat, 0.0);
        }

        for (BomEngine.BomRow row : rows) {
            double qty = row.quantity();
            hours.merge("CU", row.lbrCu() * qty, Double::sum);
            hours.merge("ASM", row.lbrAsm() * qty, Double::sum);
            hours.merge("CNT", row.lbrCnt() * qty, Double::sum);
            hours.merge("QC", row.lbrQc() * qty, Double::sum);
            hours.merge("TST", row.lbrTst() * qty, Double::sum);
            hours.merge("ENG", row.lbrEng() * qty, Double::sum);
            hours.merge("CAD", row.lbrCad() * qty, Double::sum);
        }

        return hours;
    }

    public LabourSummary costFromHours(Map<String, Double> hours, Map<String, Double> lookup) {
        Map<String, Double> resultHours = new LinkedHashMap<>();
        Map<String, Double> costs = new LinkedHashMap<>();
        Map<String, Double> rates = new LinkedHashMap<>();

        double totalHours = 0;
        double totalCost = 0;

        for (String cat : PricingEngine.LABOR_CATEGORIES) {
            double h = num(hours.get(cat));
            double rate = num(lookup.get("LBR_" + cat + "_rate"));
            double cost = h * rate;
            resultHours.put(cat, h);
            rates.put(cat, rate);
            costs.put(cat, cost);
            totalHours += h;
            totalCost += cost;
        }

        Map<String, Double> totals = new LinkedHashMap<>();
        totals.put("hours_total", totalHours);
        totals.put("cost_total", totalCost);

        return new LabourSummary(resultHours, costs, rates, totals);
    }

    public LabourSummary computeLabour(List<BomEngine.BomRow> rows, Map<String, Double> lookup) {
        return costFromHours(aggregateHoursFromBomRows(rows), lookup);
    }

    private double num(Double value) {
        return value == null ? 0.0 : value;
    }
}
